go/vt/vtgate/engine/limit.go

```diff
@@ -162,58 +154,56 @@ func (l *Limit) Inputs() []Primitive {
 	return []Primitive{l.Input}
 }
 
-//NeedsTransaction implements the Primitive interface.
+// NeedsTransaction implements the Primitive interface.
 func (l *Limit) NeedsTransaction() bool {
 	return l.Input.NeedsTransaction()
 }
 
-func (l *Limit) fetchCount(bindVars map[string]*querypb.BindVariable) (int, error) {
-	if l.Count.IsNull() {
-		return 0, nil
-	}
-
-	resolved, err := l.Count.ResolveValue(bindVars)
+func (l *Limit) getCountAndOffset(bindVars map[string]*querypb.BindVariable) (count int, offset int, err error) {
+	env := evalengine.EnvWithBindVars(bindVars)
+	count, err = getIntFrom(env, l.Count)
 	if err != nil {
-		return 0, err
+		return
 	}
-	num, err := evalengine.ToUint64(resolved)
+	offset, err = getIntFrom(env, l.Offset)
 	if err != nil {
-		return 0, err
+		return
 	}
-	count := int(num)
-	if count < 0 {
-		return 0, fmt.Errorf("requested limit is out of range: %v", num)
-	}
-	return count, nil
+	return
 }
 
-func (l *Limit) fetchOffset(bindVars map[string]*querypb.BindVariable) (int, error) {
-	if l.Offset.IsNull() {
+func getIntFrom(env *evalengine.ExpressionEnv, expr evalengine.Expr) (int, error) {
+	if expr == nil {
 		return 0, nil
 	}
-	resolved, err := l.Offset.ResolveValue(bindVars)
+	evalResult, err := expr.Evaluate(env)
 	if err != nil {
 		return 0, err
 	}
-	num, err := evalengine.ToUint64(resolved)
+	value := evalResult.Value()
+	if value.IsNull() {
+		return 0, nil
+	}
+
+	num, err := value.ToUint64()
 	if err != nil {
 		return 0, err
 	}
-	offset := int(num)
-	if offset < 0 {
+	count := int(num)
+	if count < 0 {
 		return 0, fmt.Errorf("requested limit is out of range: %v", num)
 	}
-	return offset, nil
+	return count, nil
 }
 
 func (l *Limit) description() PrimitiveDescription {
 	other := map[string]interface{}{}
 
-	if !l.Count.IsNull() {
-		other["Count"] = l.Count.Value
+	if l.Count != nil {
+		other["Count"] = evalengine.FormatExpr(l.Count)
 	}
-	if !l.Offset.IsNull() {
-		other["Offset"] = l.Offset.Value
+	if l.Offset != nil {
+		other["Offset"] = evalengine.FormatExpr(l.Offset)
```

Comment:
It would be nicer to print the raw value instead of printing type+value, like: print `1` instead of `INT64(1)`